mutation_reader.cc

```diff
@@ -2294,3 +2294,432 @@ flat_mutation_reader make_compacting_reader(flat_mutation_reader source, gc_cloc
+                            " This reader merger cannot be used for tables that have static columns"
+                            " or it would give incorrect results. The no-static-columns condition"
+                            " should have been checked before creating the reader. Please report a bug.");
+                }
+
+                if (mf->is_end_of_partition()) {
+                    _galloping_reader->reader.pop_mutation_fragment();
+                    _all_readers.erase(_galloping_reader);
+                } else {
+                    if (_reader_queue->empty(mf->position())
+                            && (_peeked_readers.empty()
+                                    || _cmp(mf->position(), _peeked_readers.front()->reader.peek_buffer().position()) < 0)) {
+                        _current_batch.push_back(_galloping_reader->reader.pop_mutation_fragment());
+
+                        return make_ready_future<mutation_fragment_batch>(_current_batch);
+                    }
+
+                    // One of the existing readers won with the galloping reader,
+                    // or there is a yet unselected reader which possibly has a smaller position.
+                    // In either case we exit the galloping mode.
+
+                    _peeked_readers.emplace_back(_galloping_reader);
+                    boost::range::push_heap(_peeked_readers, _peeked_cmp);
+                }
+            } else {
+                // See comment in `peek_reader`.
+                _halted_readers.push_back(_galloping_reader);
+                boost::range::push_heap(_halted_readers, _halted_reader_cmp);
+            }
+
+            // The galloping reader has either been removed, halted, or lost with the other readers.
+            // Proceed with the normal path.
+            _galloping_reader = {};
+            _gallop_mode_hits = 0;
+            return (*this)(timeout);
+        });
+    }
+
+public:
+    clustering_order_reader_merger(
+            schema_ptr schema, reader_permit permit,
+            streamed_mutation::forwarding fwd_sm,
+            std::unique_ptr<position_reader_queue> reader_queue)
+        : _schema(std::move(schema)), _permit(std::move(permit))
+        , _cmp(*_schema)
+        , _reader_queue(std::move(reader_queue))
+        , _peeked_cmp(*_schema)
+        , _halted_reader_cmp(*_schema)
+        , _pr_end(fwd_sm == streamed_mutation::forwarding::yes
+                        ? position_in_partition_view::before_all_clustered_rows()
+                        : position_in_partition_view::after_all_clustered_rows())
+        , _should_return_partition_end(fwd_sm == streamed_mutation::forwarding::no)
+    {
+        // The first call to `_reader_queue::pop` uses `after_all_clustered_rows`
+        // so we obtain at least one reader; we will return this reader's `partition_start`
+        // as the first fragment.
+        auto rs = _reader_queue->pop(position_in_partition_view::after_all_clustered_rows());
```

Comment:
This is confusing. A better interface would be to make the position parameter optional, when missing the next batch of readers are returned.

Reply:
But it would make the interface more complex, and `pop(std::nullopt)` would be the same as `pop({after_all_clustered_rows})`.